### PR TITLE
NSEC3 support

### PR DESCRIPTION
**Pull request by pvdrz**
This PR implements RFC 5155 including zone signing and serving of NSEC3 records.

The only unimplemented feature of the RFC is opt-out support.

### CODE REVIEW

**Review by bluejekyll — APPROVED** [review]
this looks reasonable to me. I'll be interested mostly in the validation tests...

File: crates/server/src/store/in_memory/authority.rs [hunk 1]
```diff
@@ -647,10 +666,25 @@ impl InnerInMemory {
...
+    fn secure_zone_mut(
+        &mut self,
+        origin: &LowerName,
+        dns_class: DNSClass,
+        nsec3_config: &Nsec3Config,
+    ) -> DnsSecResult<()> {
         // TODO: only call nsec_zone after adds/deletes
         // needs to be called before incrementing the soa serial, to make sure IXFR works properly
-        self.nsec_zone(origin, dns_class);
+        if nsec3_config.enable {
```
[the remark this answers is not in the document]
Comment (listochkin) [reply crates/server/src/store/in_memory/authority.rs:677]: If you have both NSEC3 and NSEC records in DNS response it is considered invalid per RFC. The way I see it there should be a three-state setting: NSEC3 only, NSEC only, and disable both.

[the remark this answers is not in the document]
Comment (pvdrz) [reply crates/server/src/store/in_memory/authority.rs:677]: I did some local experiments with this idea of a tri-state and it seems the `toml` crate cannot represent `enum`s so something like this:
```rust
pub enum NxProof {
    #[cfg(feature = "dnssec")]
    /// Use NSEC
    Nsec,
    #[cfg(feature = "dnssec")]
    /// Use NSEC3
    Nsec3,
    #[default]
    /// Do not provide non-existence proofs
    None,
}
```
cannot be used. Do you have any suggestions on how to represent this?

[the remark this answers is not in the document]
Comment (bluejekyll) [reply crates/server/src/store/in_memory/authority.rs:677]: > both NSEC3 and NSEC records in DNS response it is considered invalid per RFC

Thanks, it's been a while since I've looked at this, good to know.

>  Do you have any suggestions on how to represent this?

@pvdrz, I'm not sure what issue you're running into, here's an example of something representable in TOML: 

https://github.com/hickory-dns/hickory-dns/blob/main/crates/server/src/authority/zone_type.rs#L13-L29

**Review by marcus0x62 — CHANGES_REQUESTED** [review]

File: crates/server/src/config/mod.rs [hunk 2]
```diff
@@ -318,3 +331,38 @@ impl ZoneConfig {
...
+    pub enable: bool,
+    /// Algorithm used to create the hashed NSEC3 owner names
+    #[serde(default)]
+    pub hash_algorithm: Nsec3HashAlgorithm,
+    /// Salt used to create the hashed NSEC3 owner names
+    #[serde(default)]
+    pub salt: Vec<u8>,
+    /// Number of iterations used to create the hashed NSEC3 owner names
+    #[serde(default = "default_iterations")]
+    pub iterations: NonZeroU16,
```
[the remark this answers is not in the document]
Comment (djc) [reply crates/server/src/config/mod.rs:350]: Sounds sensible, looks like this should just be a normal `u16` instead of `NonZeroU16` instead?

File: crates/server/src/store/in_memory/authority.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pvdrz) [reply crates/server/src/store/in_memory/authority.rs]: Oh it seems I was confused about it. I pushed a commit adding an enum to handle the configuration

File: crates/server/src/store/in_memory/authority.rs [hunk 3]
```diff
@@ -722,6 +767,121 @@ impl InnerInMemory {
...
+                salt.to_vec(),
+                next_hashed_name,
+                type_bit_maps,
+            );
+
+            let name = Name::new()
+                .append_label(data_encoding::BASE32_DNSSEC.encode(hashed_name.as_ref()))
+                .unwrap()
+                .append_name(origin.borrow())
+                .unwrap();
```
Comment (marcus0x62) [inline crates/server/src/store/in_memory/authority.rs:867]: I think these unwrap calls are unsafe - a very long zone name could result in the NSEC hashes causing the total record length to be > 255 bytes, which will cause a panic here.

File: crates/server/src/store/in_memory/authority.rs [hunk 4]
```diff
@@ -1296,6 +1456,148 @@ impl Authority for InMemoryAuthority {
...
+    ) -> Result<Self::Lookup, LookupError> {
+        // Function to compute the hash of a given name.
+        let hash_name = |name: &Name| {
+            let Nsec3Config {
+                hash_algorithm,
+                ref salt,
+                iterations,
+                ..
+            } = self.nsec3_config;
+            hash_algorithm.hash(salt, name, iterations).unwrap()
```
Comment (marcus0x62) [inline crates/server/src/store/in_memory/authority.rs:1475]: There's a bug in the proto crate that masks this (I'll submit a separate PR for this,) but with that fixed, it is definitely possible to make this panic on a long zone name.

File: crates/server/src/store/in_memory/authority.rs [hunk 5]
```diff
@@ -1296,6 +1456,148 @@ impl Authority for InMemoryAuthority {
...
+
+        // Compute the hashed owner name from a given name. This is, the hash of the given name,
+        // followed by the zone name.
+        let get_hashed_name = |name: &LowerName| -> LowerName {
+            let hash = hash_name(name.borrow());
+            let label = data_encoding::BASE32_DNSSEC.encode(hash.as_ref());
+            LowerName::new(
+                &Name::new()
+                    .append_label(label)
+                    .unwrap()
```
Comment (marcus0x62) [inline crates/server/src/store/in_memory/authority.rs:1486]: This is the one I'm most concerned about - I think an attacker could send a crafted request (with a super long non-existent RR) to make the hickory server panic.

File: crates/server/src/store/in_memory/authority.rs [hunk 6]
```diff
@@ -1296,6 +1456,148 @@ impl Authority for InMemoryAuthority {
...
+        // name.
+        let find_cover = |name: &LowerName| -> Option<Arc<RecordSet>> {
+            let hash = hash_name(name.borrow());
+            let label = data_encoding::BASE32_DNSSEC.encode(hash.as_ref());
+            let owner_name = LowerName::new(
+                &Name::new()
+                    .append_label(label)
+                    .unwrap()
+                    .append_name(self.origin().borrow())
+                    .unwrap(),
```
Comment (marcus0x62) [inline crates/server/src/store/in_memory/authority.rs:1528]: Same concern as with the unwraps in nsec3_zone

**Review by marcus0x62 — CHANGES_REQUESTED** [review]

**Comment by pvdrz** [comment]
hey @marcus0x62 thanks for the review :)

Regarding tests, we have some compliance tests for NSEC3 already but currently they aren't being run with hickory because it doesn't have support for pre-signed zonefiles. I was planning on merging this PR first and then patch the compliance tests to they can be run against hickory.

File: crates/server/src/store/in_memory/authority.rs [hunk 7]
```diff
@@ -722,6 +769,122 @@ impl InnerInMemory {
...
+        // Sort by hash.
+        record_types.sort_by(|(a, _), (b, _)| a.as_ref().cmp(b.as_ref()));
+
+        // Generate an NSEC3 record for every name
+        for (i, (hashed_name, type_bit_maps)) in record_types.iter().enumerate() {
+            // Get the next hashed name following the hash order.
+            let next_index = (i + 1) % record_types.len();
+            let next_hashed_name = record_types[next_index].0.as_ref().to_vec();
+
+            let mut type_bit_maps: Vec<RecordType> = type_bit_maps.iter().copied().collect();
```
Comment (djc) [inline crates/server/src/store/in_memory/authority.rs:853]: Nit: `collect::<Vec<_>>()`, please.

File: crates/server/src/config/mod.rs [hunk 8]
```diff
@@ -318,3 +334,34 @@ impl ZoneConfig {
...
+#[cfg(feature = "dnssec")]
+/// NSEC3 related configuration.
+#[derive(Deserialize, PartialEq, Eq, Debug, Clone, Default)]
+pub struct Nsec3Config {
+    /// Algorithm used to create the hashed NSEC3 owner names
+    #[serde(default)]
+    pub hash_algorithm: Nsec3HashAlgorithm,
+    /// Salt used to create the hashed NSEC3 owner names
+    #[serde(default)]
+    pub salt: Vec<u8>,
```
Comment (djc) [inline crates/server/src/config/mod.rs:347]: Suggest we make this `Arc<[u8]>`, and adapt the `NSEC3::salt` type accordingly.

File: crates/server/src/config/mod.rs [hunk 9]
```diff
@@ -318,3 +334,34 @@ impl ZoneConfig {
...
+    #[cfg(feature = "dnssec")]
+    #[cfg_attr(feature = "dnssec", default)]
+    /// Use NSEC
+    Nsec,
+    #[cfg(feature = "dnssec")]
+    /// Use NSEC3
+    Nsec3,
+    #[cfg_attr(not(feature = "dnssec"), default)]
+    /// Do not provide non-existence proofs
+    None,
```
[the remark this answers is not in the document]
Comment (pvdrz) [reply crates/server/src/config/mod.rs:366]: > Why not model this as `Option<NxProof>`?

Two reasons:
- Because `NxProof` is shorter
- `NxProof::None` is clearly documented as the option where no non-existence proof is provided instead of having to wonder if `Option::None` means that no non-existence proof will be provided or that the default behavior will be used (which changes if the `dnssec` feature is enabled)

[the remark this answers is not in the document]
Comment (djc) [reply crates/server/src/config/mod.rs:366]: I don't find the "shorter" reason very compelling -- it's more important to me that we have a canonical way of modeling what's going on rather than fewer characters to read/write.

> * `NxProof::None` is clearly documented as the option where no non-existence proof is provided instead of having to wonder if `Option::None` means that no non-existence proof will be provided or that the default behavior will be used (which changes if the `dnssec` feature is enabled)

Is there a difference in behavior for these two cases ("not providing a non-existence proof" vs "default behavior, with dnssec disabled)?

[the remark this answers is not in the document]
Comment (japaric) [reply crates/server/src/config/mod.rs:366]: disclaimer: I'm commenting with limited context as I have not read through all the changes

In principle, I see no problem in flattening `Option<TwoVariantNxProof>` into a `ThreeVariantNxProof` enum so long we are consistent everywhere else in *not* using `Option<ThreeVariantNxProof>`. I'm less worried with the `Option<ThreeVariantNxProof>` type being allowed to exist; just like I'm not worried about it being possible to write a type like `Result<Something, Option<Error>>`

by "being consistent" I mean that the `Default` implementation of `ThreeVariantNxProof` should return the `None` variant. if the field is absent in a configuration file, serde should decode that into the `ThreeVariantNxProof::None` variant. if the flag is not passed via the command line, then assume `ThreeVariantNxProof::None` etc.

similarly, we should use the same `ThreeVariantNxProof` type regardless of whether the dnssec feature is enabled or not and we should conditionally remove the variants that make no sense when dnssec is disabled

other thing to consider is whether we want to group this with other DNSSEC related settings. perhaps at some point, it would make sense to have some `Dnssec` struct that contains `TwoVariantNxProof` (and other fields) and sits behind an `Option` when it appears in a configuration struct or something like that

[the remark this answers is not in the document]
Comment (djc) [reply crates/server/src/config/mod.rs:366]: > I'm less worried with the `Option<ThreeVariantNxProof>` type being allowed to exist;

I do think this can be an issue; `ThreeVariantNxProof` won't have such a verbose name in practice, so it's not obvious from reading that `Option<NxProof>` could have another `None` instead, and without context someone might assume that passing a non-`Option`-wrapped `NxProof` would always result in a proof being generated, when this is not the case because `NxProof` might be `NxProof::None`.

In other words, I have a strong preference towards `enum NxProofType { Nsec, Nsec3 }` with non-providers using `Option::<NxProofType>::None` to represent this in the type system, so basically I want to hear objections to why such a representation would cause more problems compared to the current one.

[the remark this answers is not in the document]
Comment (pvdrz) [reply crates/server/src/config/mod.rs:366]: >  so basically I want to hear objections to why such a representation would cause more problems compared to the current one.

Doing `enum NxProof { Nsec, Nsec3 }` and using `Option<NxProof>` everywhere means that if the `dnssec` feature is disabled you only have two choices for `NxProof`:
- Gate it behind `dnssec` as well: Which means that any item that contains `NxProof` needs to be gated through all the codebase
- Gate the two variants behind `dnssec`: Which means that you have an empty `enum` in your codebase which is not the most ergonomic thing to handle.

[the remark this answers is not in the document]
Comment (djc) [reply crates/server/src/config/mod.rs:366]: That makes sense as reasoning. Gating the existence of `NxProof` itself would make sense to me, I'm not sure how many call-sites that would touch? It seems conceptually correct.

**Comment by pvdrz** [comment]
I'm opening a new PR with this later this week as I think I'll have to redo several things from scratch instead of rebasing (which would be slower)
